crates/triangulation/src/intersection.rs

```diff
@@ -392,3 +392,323 @@ pub fn find_intersections(segments: &[point::Segment]) -> HashSet<OrderedPair> {
+    let intersections = find_intersections(&edges);
+    if intersections.is_empty() {
+        return polygon_list.to_vec();
+    }
+
+    // Create a HashMap to store intersection points for each edge
+    let mut intersections_points: HashMap<usize, Vec<(point::Coord, point::Point)>> =
+        HashMap::new();
+
+    // Process each intersection
+    for intersection in intersections {
+        let inter_points =
+            find_intersection(&edges[intersection.first()], &edges[intersection.second()]);
+
+        // Handle the intersection points based on the intersection type
+        match inter_points {
+            Intersection::PointIntersection(point) => {
+                intersections_points
+                    .entry(intersection.first())
+                    .or_default()
+                    .push((
+                        edges[intersection.first()].point_projection_factor(point),
+                        point,
+                    ));
+                intersections_points
+                    .entry(intersection.second())
+                    .or_default()
+                    .push((
+                        edges[intersection.second()].point_projection_factor(point),
+                        point,
+                    ));
+            }
+            Intersection::CollinearWithOverlap((p1, p2)) => {
+                // Handle collinear overlap case if needed
+                // to add both points to both edges
+                for point in [p1, p2] {
+                    intersections_points
+                        .entry(intersection.first())
+                        .or_default()
+                        .push((
+                            edges[intersection.first()].point_projection_factor(point),
+                            point,
+                        ));
+                    intersections_points
+                        .entry(intersection.second())
+                        .or_default()
+                        .push((
+                            edges[intersection.second()].point_projection_factor(point),
+                            point,
+                        ));
+                }
+            }
+            _ => {} // No intersection or collinear no overlap cases
+        }
+    }
+
+    // Sort and add endpoint markers for each edge's intersections
+    for (edge_idx, points) in intersections_points.iter_mut() {
+        let edge = &edges[*edge_idx];
+        points.push((-1.0, edge.top));
+        points.push((2.0, edge.bottom));
+        points.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap_or(Ordering::Equal));
+    }
+
+    // Create new polygons with intersection points
+    let mut new_polygons_list = Vec::with_capacity(polygon_list.len());
+
+    let mut polygon_shift = 0;
+
+    for polygon in polygon_list {
+        let mut new_polygon = Vec::with_capacity(polygon.len() * 2);
+        new_polygon.push(polygon[0]);
+
+        for i in 0..polygon.len() {
```

Comment:
```suggestion
        for (i, <item>) in polygon.iter().enumerate() {
```

Recommended by `cargo clippy`